cmd/thanos/receive.go

```diff
@@ -140,56 +140,96 @@ func runReceive(
 		MinBlockDuration:  model.Duration(time.Hour * 2),
 		MaxBlockDuration:  model.Duration(time.Hour * 2),
 	}
+	db := receive.NewFlushableStorage(
+		dataDir,
+		log.With(logger, "component", "tsdb"),
+		reg,
+		tsdbCfg,
+	)
 
 	localStorage := &tsdb.ReadyStorage{}
-	receiver := receive.NewWriter(log.With(logger, "component", "receive-writer"), localStorage)
 	webHandler := receive.NewHandler(log.With(logger, "component", "receive-handler"), &receive.Options{
-		Receiver:          receiver,
 		ListenAddress:     remoteWriteAddress,
 		Registry:          reg,
-		ReadyStorage:      localStorage,
 		Endpoint:          endpoint,
 		TenantHeader:      tenantHeader,
 		ReplicaHeader:     replicaHeader,
 		ReplicationFactor: replicationFactor,
 	})
 
+	confContentYaml, err := objStoreConfig.Content()
+	if err != nil {
+		return err
+	}
+	upload := true
+	if len(confContentYaml) == 0 {
+		level.Info(logger).Log("msg", "No supported bucket was configured, uploads will be disabled")
+		upload = false
+	}
+
 	// Start all components while we wait for TSDB to open but only load
 	// initial config and mark ourselves as ready after it completed.
-	dbOpen := make(chan struct{})
+
+	// dbReady signals when TSDB is ready and the Store gRPC server can start.
+	dbReady := make(chan struct{})
+	// updateDB signals when TSDB needs to be flushed and updated.
+	updateDB := make(chan struct{})
+	// uploadC signals when new blocks should be uploaded.
+	uploadC := make(chan struct{})
+	// uploadDone signals when uploading has finished.
+	uploadDone := make(chan struct{})
 	level.Debug(logger).Log("msg", "setting up tsdb")
 	{
 		// TSDB.
 		cancel := make(chan struct{})
-		g.Add(
-			func() error {
-				level.Info(logger).Log("msg", "starting TSDB ...")
-				db, err := tsdb.Open(
-					dataDir,
-					log.With(logger, "component", "tsdb"),
-					reg,
-					tsdbCfg,
-				)
-				if err != nil {
-					close(dbOpen)
-					return fmt.Errorf("opening storage failed: %s", err)
-				}
-				level.Info(logger).Log("msg", "tsdb started")
-
-				startTimeMargin := int64(2 * time.Duration(tsdbCfg.MinBlockDuration).Seconds() * 1000)
-				localStorage.Set(db, startTimeMargin)
-				webHandler.StorageReady()
-				level.Info(logger).Log("msg", "server is ready to receive web requests.")
-				close(dbOpen)
-				<-cancel
-				return nil
-			},
-			func(err error) {
-				if err := localStorage.Close(); err != nil {
-					level.Error(logger).Log("msg", "error stopping storage", "err", err)
+		// Before actually starting, we need to make sure
+		// the WAL is flushed.
+		startTimeMargin := int64(2 * time.Duration(tsdbCfg.MinBlockDuration).Seconds() * 1000)
+		if err := db.Open(); err != nil {
```

Comment:
Yes we could do that. I was worried that it would be confusing if flush left an open DB. Does you think it would be more expected for Flush to leave a running DB?

Reply:
Yes we could do that. I was worried that it would be confusing if flush left an open DB. Does you think it would be more expected for Flush to leave a running DB?

Reply:
I think it is most logical that flush should: `open the db -> flush -> close the db`

Reply:
there are two  issues I see with that:

1. `f.Open()` returns an error, so we need to error check `f.Get()`
2. if flush expects a closed DB, then the caller has to call `close` before flushing.

Both of these mean more work for the caller. For example:

```go
// Using DB
if err := f.Close(); err != nil {
}
if err := f.Flush(); err != nil {
}
db, err := f.Get()
if err != nil {
}
// Using DB
```
instead of simply:
```go
// Using DB
if err := f.Flush(); err != nil {
}
db := f.Get()
// Using DB
```

Reply:
this can also happen internally, no?
```
if !db.closed {
	db.Close()
}
.
.
```

Reply:
Sure I just want to be clear about the API. I think it’s confusing if Flush leaves an open DB if the original state was closed. It may be clearest to say `Flush will always leave the storage in the same state it was in: open storage will stay open; closed storage will remain closed.`

Reply:
yep , that is a great idea, but can you imagine a case where the storage would be open when running a flush?

Reply:
The DB is open immediately before every flush (we are actively writing samples until we get the reshard signal and need to flush), someone has to close it, either the caller, or the flush method itself

Reply:
hm yeah, maybe now with the new tsdb.Flush method can just have just use it directly and not need the separate `FlushableStorage`

Reply:
Yes that’s what I was driving at earlier. It will reduce the code here, though we’ll have some more boilerplate in the caller since it will have to open a new DB after flushing. This type is convenient because we can have one single reference that takes care of re-opening itself after flushing if it was already open and the only thing the caller has to do is call `Open` once at the very beginning with all the args and then `Flush`